### ISSUE TITLE
[RFC] API for pluggable work queue and distributing work to remote workers

### ISSUE THREAD

**Issue by adrianke77** [backstage/backstage#6816]
## Need
Several parts of Backstage involve compute intensive work that would be better handled away from the process handling API requests, such as
 - Entity catalog ingestion
 - Scaffolding
 - Techdocs rendering
It would be good to have such handled by workers placed elsewhere from the API server with a queue implementation, however at this time such a setup is not supported.

## Proposal
Backstage could have a API that allows externally implemented queues to handle tracking and distributing tasks to asynchronous workers.

The position of the queue would be a middleman between requests from plugins to perform some task and for other copies of the plugins running on different nodes to handle that task. The setup would be that plugins declare they can handle certain asynchronous work (scaffolding, Techdocs rendering, etc) by registering some lambdas as part of their initialisation, and plugins can request work to be done by passing a task description JSON, specifying the plugin and lambda name they want to handle the task, along with arguments and data required for the task. 

So in the default case of no queue being used, tasks are passed from requesters to handlers directly, such as performing a setTimeout(<handler lambda with passed args/data>,0). When a queue is additionally implemented, the tasks are instead passed to the queue and enqueued, and the queue implementation then handles tracking and distributing the tasks to available handlers. 

Such work should not require any response signal when done; polling by requesters to to find if work has finished could be used if needed, such as Techdocs rendering leaving the new eTag in the docs storage location to track the corresponding commit of the last render.

Remote workers would be other instances of Backstage elsewhere that only handle incoming task requests but not API requests. The external implementation around the queue would be responsible for communication between Backstage requester and worker instances. This would allow a lot of flexibility to manage distributed use of resources, not only from API servers to workers but also from worker to worker.

Scheduled/cadenced tasks such as entity catalog background ingestion would have their expected execution time or interval specified in the task description payload, to be appropriately handled by the queue implementation. With no queue these are handled directly as timeouts etc. 

## Alternatives
Tasks could be handled by instancing worker threads within Backstage, however this limits available resources for workers to the same physical system/node the API server lives in. 

## Risks
This could conflict with ongoing work for scaffolders or other similar task handling. However some refactoring would probably allow current/future implementations of scaffolders, ingestion etc to work with the above proposed pattern.

**Comment by tragiclifestories** [comment]
Some related previous discussion: https://github.com/backstage/backstage/issues/3746

**Comment by iamEAP** [comment]
I would add that the Search Collation/Indexing process would be a strong candidate for this.  I'm less familiar with scaffolder/catalog ingestion generally, but Search Indexing is unique in that it needs to phone back home to specific plugin backends to collate documents (e.g. Catalog API to read entities, TechDocs API to read pages, etc), which would likely necessitate a solution to #5210.

With that in mind, I would also wonder if any considerations need to be made in cases where, for example, the Catalog backend is deployed distinct from the TechDocs backend as well as distinct from the Search backend...  And then this queue processing pool would _also_ be deployed distinctly.  Would the a Search Indexing lambda have enough context to resolve the location of the Catalog API or the TechDocs API?

**Comment by freben** [comment]
Architecturally, both the catalog and the scaffolder are heavily built on using the database as a synchronisation/coordination primitive. At least for those two, an event bus _for the purposes of coordination or distribution of load_ is not of any particular urgency or interest at least short term since that architecture is unlikely to be revamped right now.

However, they are very interested in both consuming and producing structured events. For example, the catalog is longing to be able to instantly react to SCM events and to communicate to the outside world when entities have updates etc. But that problem space / framing is a bit different to what this issue mentions, it seems to me.

On the topic of separating load from the API handling, that is indeed desirable and we would like to find a way of addressing that soon. One way of doing so would be to simply leverage service workers to put that load on a separate dedicated thread. Or to use similar fork-like solutions for separation of concerns.

**Comment by freben** [comment]
@iamEs me plugin to plugin comms would still be done using clients that use the standard [pluggable resolution mechanism](https://github.com/backstage/backstage/blob/master/packages/backend-common/src/discovery/types.ts#L29).

**Comment by lowjoel** [comment]
> Architecturally, both the catalog and the scaffolder are heavily built on using the database as a synchronisation/coordination primitive. At least for those two, an event bus _for the purposes of coordination or distribution of load_ is not of any particular urgency or interest at least short term since that architecture is unlikely to be revamped right now.

Hmm, I think it's a slightly different ask. This isn't quite an event bus - this is just a way to delegate work to a different compute node from the one that's serving API traffic. A NPM package that does this today is https://github.com/actionhero/node-resque -- not saying we need to use this, but resque _could_ be a candidate implementation for such a primitive.

The use of the database is causing quite a bit of load - and causing problems with scaling (especially since it's a relational database). In theory, it should be possible to extract the database synchronisation work into a possible work queue implementation, so that by default, nothing actually changes (well - some minor API changes may be needed, not investigated yet.) At the same time, for installations with significant database load (like ours), we can swap it out for a different implementation that uses something else and push the compute elsewhere from the API server.

> One way of doing so would be to simply leverage service workers to put that load on a separate dedicated thread. Or to use similar fork-like solutions for separation of concerns.

Definitely, even though I'm somewhat against this - since this prioritises vertically scaling compute, rather than horizontally scaling them. This makes cost optimisation much harder for most cloud (and Kubernetes) workloads.

@freben does the ability to continue using the database as a work queue implementation help change your mind about the desirability of such an implementation? I have not looked into the code, but experience with distributed systems so far indicate that this particular primitive is incredibly useful.

**Comment by freben** [comment]
> > Architecturally, both the catalog and the scaffolder are heavily built on using the database as a synchronisation/coordination primitive. At least for those two, an event bus _for the purposes of coordination or distribution of load_ is not of any particular urgency or interest at least short term since that architecture is unlikely to be revamped right now.
> 
> Hmm, I think it's a slightly different ask. This isn't quite an event bus - this is just a way to delegate work to a different compute node from the one that's serving API traffic. A NPM package that does this today is https://github.com/actionhero/node-resque -- not saying we need to use this, but resque _could_ be a candidate implementation for such a primitive.

That looks like a great library! I'll just add here that one guiding principle we've had so far is that we have really tried to not introduce any more required moving parts in a deployment, than the database. Adding redis as one more piece definitely has operational overhead, especially for smaller adopters. I'm not saying no to the idea, but rather that we are hesitant to doing something like this and will probably require a bit of persuasion to taking the step :)

> The use of the database is causing quite a bit of load - and causing problems with scaling (especially since it's a relational database). In theory, it should be possible to extract the database synchronisation work into a possible work queue implementation, so that by default, nothing actually changes (well - some minor API changes may be needed, not investigated yet.) At the same time, for installations with significant database load (like ours), we can swap it out for a different implementation that uses something else and push the compute elsewhere from the API server.

The way that processing is triggered, is currently via timestamps that live inside the database, on each `refresh_state` row. Even if coordination happened through a bus, that timestamp would have to be bumped, and the state row contents transferred to the bus.

Essentially there's no "push" for work. Each worker "asks" for work by trying to claim rows in the database as their own for a certain period of time. It's a combination of work stealing and mutual exclusion / locking.

To be clear, the database load you are seeing is most likely due to excessive, unnecessary write + stitch work, which was just [recently addressed](https://github.com/backstage/backstage/pull/6827), so it may be the case that you will soon see drastically reduced load anyway. In addition to that, we are this week looking into instrumenting and improving the database interactions even more - there's a lot of room for improvement since the recent rewrite was just a first iteration. And on a longer time scale, we'll look into making these things event driven as well, for example reacting to webhooks from your upstreams so as not to perform any unnecessary work at all. Some ground work for that is [starting to take form](https://github.com/backstage/backstage/pull/6441) right now.

> > One way of doing so would be to simply leverage service workers to put that load on a separate dedicated thread. Or to use similar fork-like solutions for separation of concerns.
> 
> Definitely, even though I'm somewhat against this - since this prioritises vertically scaling compute, rather than horizontally scaling them. This makes cost optimisation much harder for most cloud (and Kubernetes) workloads.

I think if we went with the worker method I mentioned, it would not be targeted at scaling vertically as such. It is rather because how the router and worker code are currently in the same package - you would just run the worker code in a separate thread only to not have performance interactions with the api thread. Then you would scale that pair basically just like before.

Or, you could introduce the ability to run the backend in one of two (single threaded) modes - api mode or worker mode (just a config flag or so), and deploy and scale those separately. The other alternative would be to break apart the catalog-backend package into two actual separate packages. Either way, for all of those solutions, I was suggesting that maybe they would interact only via the database anyway.

> @freben does the ability to continue using the database as a work queue implementation help change your mind about the desirability of such an implementation? I have not looked into the code, but experience with distributed systems so far indicate that this particular primitive is incredibly useful.

Just want to double check - what did you mean with being able to continue using the database for that? Did you mean that if we disregard the catalog use case, to still introduce this kind of capability into the framework for other purposes than the catalog? If so, what are the primary use cases to target in the near term?

---

Let me just also add here that I absolutely do not want to sound dismissive. It's hard to get across the right way in writing! I think this is an awesome architectural discussion to have, and I really really wish we were in a position to pull things apart a bit more and lean on this kind of primitive, so your ideas are sound and come from a good place. I'm more thinking out loud about practicalities and timeframes etc.

**Comment by lowjoel** [comment]
> That looks like a great library! I'll just add here that one guiding principle we've had so far is that we have really tried to not introduce any more required moving parts in a deployment, than the database. Adding redis as one more piece definitely has operational overhead, especially for smaller adopters. I'm not saying no to the idea, but rather that we are hesitant to doing something like this and will probably require a bit of persuasion to taking the step :)

Yeah, and I fully support that goal. Ease of onboarding helps keep the cost of adopting Backstage low. Hence my suggestion to abstract this primitive away, and hide it behind an interface. @adrianke77 didn't really go into detail, but I can imagine as a rough sketch:

```typescript
interface Task {
  operation: string; // Or something that identifies the work done, could be scoped per-plugin, not sure yet
  args: JsonObject;
}
interface WorkQueue {
  do(task: Task);
  doLater(task: Task, at Date);
}
```

So the existing code that needs this work done can use this interface. We can then push work onto the WorkQueue implementation - could be a database (as it is today in Catalog), or a Redis (like Resque). The opposite end, the consumer:

```typescript
interface TaskFunction {
  (args: JsonObject): void;
}
interface QueueProcessor {
  registerOperation(operation: string, f: TaskFunction);
}
```

Then plugins would just need to call `registerOperation` on some the QueueProcessor provided.

So now let's implement a WorkQueueProcessor:

```typescript
class DatabaseWorkQueueProcessor implements WorkQueue, QueueProcessor {
  registerOperation(operation: string, f: TaskFunction) {
    // keep in a map, know that all Tasks with this operation, call f(args)
    this.operations[operation] = f;
  }

  do(operation: string, args: JsonObject) {
    // insert into table with operation, and arguments serialised
  }

  doLater(...) { ... } // same as do, just with a timestamp later

  run() {
    // This could be your "worker mode" entry point
    while (true) {
      let work: Task = workSteal(); // select from work queue table, update row with lease timeout
      this.operations[work.operation](work.args)
    }
  }
}
```

The exact same pattern can be replaced with resque -- with probably more specialised implementations than what we have for a relational database (optimised for Redis, or something). I think that's the beauty of this architecture primitive! 😁 The existing catalog background worker can fit _right into_ this design. But we have now generalised this to work for _any_ plugin with a similar demand (caveats for things like plugin isolation etc.)

> it may be the case that you will soon see drastically reduced load anyway. In addition to that, we are this week looking into instrumenting and improving the database interactions even more 

Good to hear! However, I'm bringing this up because I think there is a much more general set of problems that can be solved with this primitive 😄  This single primitive allows so many kinds of work to be abstracted on top of.

> you would just run the worker code in a separate thread only to not have performance interactions with the api thread. Then you would scale that pair basically just like before.

Yeah, and that's a little wastage there - because now I can scale both by the same amount, not by the demands of each. CPU is not a problem, but I think memory is. Since cloud providers usually give memory:CPU in fixed ratios, being able to scale exactly the kind of work that's needed would be helpful in saving costs in the long run. That's what I mean by vertically scaling - I now give each copy of backstage-backend twice as much compute to utilise (4vCPU for example, but also 2x the memory), when I could split it up to two pods (in Kubernetes) of 2vCPU and 1x the memory. Now if I have more API requests, I scale the API server out. If I have more async work, I scale the workers out. Instead of scaling up a pair - which doubles my resources and leaves one half unused, I scale up the part which needs more. Assuming I give same memory and CPU to each, for the same total amount of resources I can now run 3 copies of background worker+1 api server, where in the paired scenario I would have 2 workers+2 api servers (with API server resources left idle.)

Granted this is only possible with more configuration on the user's end, where you specify "api-only mode", "worker-only mode" and "both mode" -- but for larger installs this would scale much better. We could default to "both mode" and let users decide how to roll out. Today we already deploy the backend as "frontend mode" and "backend mode" - because of the way the frontend bundle is deployed 😄 

> what did you mean with being able to continue using the database for that? Did you mean that if we disregard the catalog use case, to still introduce this kind of capability into the framework for other purposes than the catalog? If so, what are the primary use cases to target in the near term?

Yes - @adrianke77's post suggested:
- Entity catalog ingestion
- Scaffolding
- Techdocs rendering

@iamEAP suggested for search indexing.

In full disclosure - by "techdocs rendering" - we did an analysis of the options available to us at this point and figured out that from a security/audit/access control perspective, keeping the TechDocs to be a pull model works better. I know the recommended workflow is to build in CI - but that brings a host of other problems regarding access control, auditing, key rotation, abuse; and so keeping the pull model where Backstage maintains access to code repositories is a more palatable solution.

What we're planning to implement is to utilise this pluggable work queue for rendering outdated docs. We can scale the workers to cope with demand, while not bogging down the API server. This gives us the central control of access to our code repositories, and keeping the cognitive overhead of individual teams down (they just tell us where their docs are; no need to fumble with permissions and commands and things like that -- which our environment would require at the moment.)

> I'm more thinking out loud about practicalities and timeframes etc.

For sure! This is not a short-term ask. It'll take time to properly sketch out the design and come up with an implementation. It's effort on the order of weeks/months. I strongly believe that such a primitive would be incredibly useful at building Backstage to work at much larger scales, and solves a ton of other problems too, hence we wanted to get this discussion started early.

**Comment by Rugvip** [comment]
@lowjoel A PostgreSQL work queue implemented with `SKIP LOCKED` is quite performant. I think it's best to have some numbers at hand before optimizing that part and focus more on getting worker processes and nodes in as part of the architecture.

**Comment by x3ro** [comment]
I would like to re-open the discussion on this RFC. If I understand correctly, what's being proposed here is similar to what we have been contemplating, mainly for the purpose of rate-limiting. A mechanism similar to what's currently implemented for the TaskScheduler, but by having a table which contains a list of `Tasks` to run.

Some background: We are working with a large (10k repos) GitHub organization, and we have adopted the recently included `events` module in order to receive webhooks from Github and update catalog entries in response. We have entirely disabled refresh and scanning, because having them enabled exhausts our API limits way too quickly. 

A task queue would, I think, enable us to implement the following mechanisms:

* Never run more than 100 instances of task type `github-catalog-update` in an hour
* Never update the same entity more than once every 10 minutes
* Deprioritize TechDocs updates for an entity that has already been updated three times this hour

etc. 

You probably get the gist, essentially we're interesting in rate limiting for tasks triggered either internally or through external sources. We've implemented a prototype based on [pg-boss](https://github.com/timgit/pg-boss) but are not yet using it in production (I will figure out if I can share the source).

Long story short, I'm wondering if there is interest for such a feature in Backstage. If I understand correctly, this could also potentially act as a replacement for the TaskScheduler 🤔

**Comment by andrewthauer** [comment]
We've been also contemplating some sort of reusable background task runner. However our use case has nothing to do with the catalog ingestion. I do think having some sort of official task queuing API available in Backstage would benefit other plugins requiring a queue and not a just a cron scheduler. This would help create some consistency for this use case which comes up decently often imo.

**Comment by andrewthauer** [comment]
I'm going to re-open this as I feel like it's worth discussing further the need for an official plugin API for background tasks.

**Comment by jmealo** [comment]
I'm interested in taking a swing at this. I'm mostly interested in adding this for k8s crons. I took a look at the available interface components in Backstage and the addition of a calendar or timeline view would be required for my ideal interface.

**Inputs:**
- CatalogInfo
- ? CronTab
- ? K8s (query/api or labels?)

**Proposed UI:** 
- Timeline viewer (react-crono?)
- Manually Invoke/Re-run jobs?

**Schema considerations:**
Declare business and service dependencies on a given job. Allow severity/escalation/run-book/business impact to be declared to power alerts.

**Interface Views:**
- Diaster recovery run book
- historic timeline
- status page

Can we expose these as templates so folks can tweak declaratively?

**Remaining questions:**
I'm likely going to expose the UI to view logs and manually re-run failed jobs/manually  dispatch them using GitHub actions to reduce complexity for an MVP and have good audit logs.

- Do we want these views to be read only?
- What are the security considerations?
- What are folks must haves for an MVP?

🕜 I just got started with backstage last week so pardon my spitballing here. Please correct  me or fill in any gaps as needed. :)
